drivers/hwinfo/hwinfo_nrf.c

```diff
@@ -25,17 +25,34 @@ struct nrf_uid {
 ssize_t z_impl_hwinfo_get_device_id(uint8_t *buffer, size_t length)
 {
 	struct nrf_uid dev_id;
-	uint32_t deviceid[2];
+	uint32_t buf[2];
 
+#if NRF_FICR_HAS_DEVICE_ID || NRF_FICR_HAS_INFO_DEVICE_ID
+	/* DEVICEID is accessible, use this */
 #if defined(CONFIG_TRUSTED_EXECUTION_NONSECURE) && defined(NRF_FICR_S)
-	soc_secure_read_deviceid(deviceid);
+	soc_secure_read_deviceid(buf);
 #else
-	deviceid[0] = nrf_ficr_deviceid_get(NRF_FICR, 0);
-	deviceid[1] = nrf_ficr_deviceid_get(NRF_FICR, 1);
+	buf[0] = nrf_ficr_deviceid_get(NRF_FICR, 0);
+	buf[1] = nrf_ficr_deviceid_get(NRF_FICR, 1);
+#endif
+#elif NRF_FICR_HAS_DEVICE_ADDR || NRF_FICR_HAS_BLE_ADDR
+	/* DEVICEID is not accessible, use device/ble address instead.
+	 * Assume that it is always accessible from the non-secure image.
+	 */
+	buf[0] = nrf_ficr_deviceaddr_get(NRF_FICR, 0);
+	buf[1] = nrf_ficr_deviceaddr_get(NRF_FICR, 1);
+
+	/* Assume that ER and IR are available whenever deviceaddr is.
+	 * Use the LSBytes from ER and IR to complete the device id.
+	 */
+	buf[1] |= ((nrf_ficr_er_get(NRF_FICR, 0) & 0xFF) << 16);
+	buf[1] |= ((nrf_ficr_ir_get(NRF_FICR, 0) & 0xFF) << 24);
```

Comment:
Outer parenthesis are not needed

Reply:
Right, but I actually prefer them for clarity.

Reply:
We do not do that, it is bad overparenthesizing and environmental pollution.

Reply:
OK but let me do a cleanup round after this is merged.